pkg/networkmanager/firewall.jsx

```diff
@@ -0,0 +1,265 @@
+    constructor() {
+        super();
+
+        this.state = {
+            selected: new Set()
+        };
+
+        this.onToggleService = this.onToggleService.bind(this);
+    }
+
+    componentDidMount() {
+        firewall.getAvailableServices()
+            .then(services => this.setState({ services: services }));
+    }
+
+    onToggleService(event) {
+        var service = event.target.id;
+        var enabled = event.target.checked;
+
+        this.setState(oldState => {
+            var selected = new Set(oldState.selected);
+
+            if (enabled)
+                selected.add(service)
+            else
+                selected.delete(service);
+
+            this.props.selectionChanged(selected);
+
+            return {
+                selected: selected
+            };
+        });
+    }
+
+    render() {
+        if (!this.state.services) {
+            return (
+                <div className="modal-body">
+                    <div className="spinner spinner-lg" />
+                </div>
+            );
+        }
+
+        return (
+            <div id="add-services-dialog" className="modal-body">
+                <ul className="list-group dialog-list-ct">
+                    {
+                        this.state.services.map(s => (
+                            <li className="list-group-item">
+                                <label>
+                                    <input id={s.id}
+                                           type="checkbox"
+                                           checked={this.state.selected.has(s.id)}
+                                           onClick={this.onToggleService} />
+                                    &nbsp;
+                                    <span>{s.name}</span>
+                                </label>
+                            </li>
+                        ))
+                    }
+                </ul>
+            </div>
+        );
+    }
+}
+
+export class Firewall extends React.Component {
+    constructor() {
+        super();
+
+        this.state = {
+            installed: firewall.installed,
+            enabled: firewall.enabled,
+            services: firewall.services,
+            enabledServices: firewall.enabledServices
+        };
+
+        this.onFirewallChanged = this.onFirewallChanged.bind(this);
+        this.onAddServices = this.onAddServices.bind(this);
+        this.onRemoveService = this.onRemoveService.bind(this);
+    }
+
+    onFirewallChanged() {
+        this.setState({
+            installed: firewall.installed,
+            enabled: firewall.enabled,
+            services: firewall.services,
+            enabledServices: firewall.enabledServices
+        });
+    }
+
+    onSwitchChanged(value) {
+        if (value)
+            firewall.enable();
+        else
+            firewall.disable();
+    }
+
+    onAddServices() {
+        var services = [...this.state.services].map(id => this.state.services[id]);
```

Comment:
Why does this need a copy of state.services? The .map already creates a new object.

Reply:
I agree, this is very unfortunate. `this.state.services` is a `Set`, which doesn't have `map` defined.